bigquery/google/cloud/bigquery/client.py

```diff
@@ -2080,6 +2083,54 @@ def insert_rows(self, table, rows, selected_fields=None, **kwargs):
 
         return self.insert_rows_json(table, json_rows, **kwargs)
 
+    def insert_rows_from_dataframe(
+        self, table, dataframe, selected_fields=None, chunk_size=500, **kwargs
+    ):
+        """Insert rows into a table from a dataframe via the streaming API.
+
+        Args:
+            table (Union[ \
+                :class:`~google.cloud.bigquery.table.Table`, \
+                :class:`~google.cloud.bigquery.table.TableReference`, \
+                str, \
+            ]):
+                The destination table for the row data, or a reference to it.
+            dataframe (pandas.DataFrame):
+                A :class:`~pandas.DataFrame` containing the data to load.
+            selected_fields (Sequence[ \
+                :class:`~google.cloud.bigquery.schema.SchemaField`, \
+            ]):
+                The fields to return. Required if ``table`` is a
+                :class:`~google.cloud.bigquery.table.TableReference`.
+            chunk_size (int):
+                The number of rows to stream in a single chunk. Must be positive.
+            kwargs (dict):
+                Keyword arguments to
+                :meth:`~google.cloud.bigquery.client.Client.insert_rows_json`.
+
+        Returns:
+            Sequence[Sequence[Mappings]]:
+                A list with insert errors for each insert chunk. Each element
+                is a list containing one mapping per row with insert errors:
+                the "index" key identifies the row, and the "errors" key
+                contains a list of the mappings describing one or more problems
+                with the row.
+
+        Raises:
+            ValueError: if table's schema is not set
+        """
+        insert_results = []
+
+        chunk_count = int(math.ceil(len(dataframe) / chunk_size))
+        rows_iter = (row._asdict() for row in dataframe.itertuples(index=False))
```

Comment:
Is `_asdict()` because `itertuples()` returns `NamedTuples`? Do we know what versions of Pandas support this? Do we need to update the optional dependency in `setup.py`?

Reply:
The `itertuples()` method was added in Pandas `0.7.1` ([release notes](https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.7.1.html?highlight=itertuples)),  and support for namedtuples was [added in v0.17.1](https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.17.1.html?highlight=itertuples#enhancements). We have the `pandas` version pinned to `>=0.17.1` in `setup.py`.

With that said, looking at the source reveals an [unpleasant trap](https://github.com/pandas-dev/pandas/blob/v0.25.1/pandas/core/frame.py#L969-L972) - if the number of columns is larger than 255, a _regular_ tuple is returned, and there is no `_asdict()` method on it.

I will rewrite this part to construct a dict from the tuple manually.

Reply:
Well go figure, `itertuples()` has a bug has a bug in Pandas `0.25.1` (but not reproducible in `0.24.2`):
```py
>>> import pandas
>>> pandas.__version__
'0.25.1'
>>> df = pandas.DataFrame([{"foo_{}".format(i): "bar_{}".format(i) for i in range(255)}])
>>> df.itertuples(index=False)
...
SyntaxError: more than 255 arguments
```
They introduced it [here](https://github.com/pandas-dev/pandas/commit/81f5c0164e6cb26cc8fea00ebc22f4fe93771ff6#diff-1e79abbbdd150d4771b91ea60a4e1cc7R903) by removing the try-except block around the namedtuple creation.

Will enforce plain tuples by passing `name=None` to `itertuples()`, and report the issue to Pandas maintainers.